rfcs/0164-bootspecv2.md

```diff
@@ -0,0 +1,300 @@
+  //  - Each entry in the toplevel "org.nixos.specialisation.v2" object represents a specialisation.
+  //  - In order for the top-level document to be a valid v2 bootspec, each specialisation must have a valid "org.nixos.bootspec.v2" key whose value conforms to the same schema as the toplevel "org.nixos.bootspec.v2" object.
+  //  - The behaviour of nested specialisations (i.e. entries in "org.nixos.specialisation.v2" which themselves contain the "org.nixos.specialisation.v2" key) is not defined.
+  //  - In particular, there is no expectation that such nested specialisations will be handled by consumers of bootspec documents.
+  //  - Each specialisation document may contain arbitrary further keys (extensions), like the top-level document.
+  //  - The semantics of these should be the same as when these keys are used at the top level, but only apply for the given specialisation.
+  "org.nixos.specialisation.v2": {
+    // Each key in this object corresponds to a specialisation as defined by the `specialisation.<name>` NixOS option.
+    "<name>": {
+      "org.nixos.bootspec.v2": {
+        // See above
+      }
+    }
+  },
+}
+```
+
+An *optional* field means: a field that is either missing or present, but **never `null`**.
+
+### Risks
+[risks]: #risks
+
+- Some of the bootloader backends are quite complicated, and in many cases have
+  inadequate tests. We could accidentally break corner cases.
+- The bootloader backends are inherently a weak point for NixOS, as it is our
+  last option for rolling back. We cannot roll back a broken bootloader. This
+  and the previous point are risks, but also help demonstrate the value of
+  reducing the amount of code and complexity in the generator.
+
+### Milestones
+[milestones]: #milestones
+
+- Update Bootspec with the version 2 of that specification.
+- Implement changes inside at least one bootloader backend.
+
+# FAQ
+[faq]: #faq
+
+Familiarize yourself with Bootspec v1 which may already contain answers.
+
+## How will I be able to express complicated logic for initrd secrets, e.g. dynamic secrets?
+
+Some users may have used the appender script to provide dynamic logic that
+provides a secret at activation time rather than storing it on the long run.
+
+While this usecase is interesting, it is very advanced and the footgun that
+initrd secrets represent cannot be made up by our only goal of supporting that
+usecase.
+
+Authors may propose to think about how a filesystem could implement the dynamic
+fetching at activation time, e.g. a FUSE secretfs that will dynamically query
+the secret engine for a secret and make it available for short time.
+
+If any logic requires user interaction, it is preferable to invest in a custom
+bootloader installer logic and use the static fields to refer to secrets that
+will be requested. We never defined "path" on purpose as it could be a network
+path or an application-specific path.
+
+It could even be a UEFI device path!
```

Comment:
I don't think this is a good idea, and opens up a lot of risky potential security issues where a security-sensitive value may be mishandled. I think it would be much more preferable to have a strong definition of what these values are and how they are to be interpreted. I think it would be better if different behavior was implemented by extensions which also had strong definitions of how to handle the values.

Putting undefined at the trust root smells bad to me.

Reply:
Hmm, so with your idea, would we be doing formalized extensions of `initrdSecrets` and let people experiment with their own definitions of `initrdSecrets` ?

Reply:
And we would strongly say those are paths on the local system, right?

Reply:
My preference would be `org.nixos.bootspec.v2` have no mention of `initrdSecrets`, and the definition l interface is in a separate extension like `org.nixos.initrd-secrets.v1` (which could be defined in this same RFC) or something.

Reply:
OK, this makes sense to me.

Reply:
I actually also like this idea

Reply:
I was actually loathe to include initrd secrets in v1, but I don't think we'd thought about Extensions yet, and also didn't think to clean it up after extensions were added.